### ISSUE TITLE
Enable mutliple grants not working as expected for "snowflake_database_grant" & "snowflake_view_grant"

### ISSUE THREAD

**Issue by amit0904** [Snowflake-Labs/terraform-provider-snowflake#1459]
**Provider Version**

0.53.0

**Terraform Version**

1.3.6

**Describe the bug**

For resource "snowflake_database_grant", provider removes views & tables from share database even when enable_multiple_grants is set to true.

**Expected behavior**

Existing views and tables should not be removed

**Code samples and commands**

```
resource "snowflake_database_grant" "db_grant" {
  provider                       = snowflake.sysadmin
  database_name            = local.database_name
  enable_multiple_grants = true

  privilege = var.db_privilege
  roles     = [snowflake_role.this.name]

  with_grant_option = false
}

resource "snowflake_view_grant" "this" {
  provider      = snowflake.sysadmin
  for_each      = local.view_list
  database_name = var.database_name
  schema_name   = var.schema_name
  view_name     = each.value

  privilege = var.privilege

  roles  = var.roles

  depends_on = [
    data.snowflake_views.this
  ]

  enable_multiple_grants = true
  with_grant_option      = false
}
```

**Output**
```
  module.roles.module.read-only-roles.module.outbound_db_read_only_grants.module.grant_on_all_views["SCHEMA_NAME"].snowflake_view_grant.this["VIEW_NAME"] will be updated in-place
  ~ resource "snowflake_view_grant" "this" {
        id                     = "OUTBOUND|SCHEMA_NAME|VIEW_NAME|SELECT|OUTBOUND_DB_READ_ONLY|false"
      ~ shares                 = [
          - "OUTBOUND",
        ]
        # (8 unchanged attributes hidden)
    }
```

**Comment by amit0904** [comment]
Hi @brandocomando ,

If you look at the code snippet again you will see that I am doing the same. There are no shares mentioned and I am just apply grant to the roles but then also it drop/removes schema from the share.

**Comment by Anatoli-deBRADKE** [comment]
I observe a similar behavior on my side. 
All the database USAGE privilege granted to my shares (Declared outside my terraform code) are removed. The enable_multiple_grants parameter does not seem to work as expected. 

**Input**

```terraform
resource "snowflake_database" "historization" {
  provider = snowflake.sys_admin

  name = upper("HISTORIZATION_DATABASE_${var.common_vars.env}")
}

resource "snowflake_role" "historization" {
  provider = snowflake.security_admin

  name = upper("HISTORIZATION_ROLE_${var.common_vars.env}")
}

resource "snowflake_database_grant" "historization_usage" {
  provider = snowflake.security_admin

  depends_on = [
    snowflake_database.historization,
  ]

  database_name  = snowflake_database.historization.name
  privilege  = "USAGE"
  with_grant_option  = true
  enable_multiple_grants = true

  roles = [
    snowflake_role.historization.name,
  ]
}
```

**Comment by sfc-gh-swinkler** [comment]
So the problem with the existing grant resources is that they accept a list of shares / roles. During the read Terraform runs the command `SHOW GRANTS OF <OBJECT_TYPE> <OBJECT_NAME>` , and then uses this output to detect if configuration drift has occurred on the roles / shares that have this privilege. In my opinion, this is a backwards and inherently flawed way of looking at the problem because grants can be created through multiple different means, and you won't have any idea what roles and shares your resource should be managing. The `enable_multiple_grants` attribute is an attempt to solve the problem by disabling the normal way that Terraform works and simply not performing an update if the list has changed. I think what we should do instead is turn the problem on its head, and create all new resources that only manage the grant for a specific role / database role / share, which more closely matches the API. I have written about a proposed solution here: https://github.com/Snowflake-Labs/terraform-provider-snowflake/discussions/1470

I don't know if we should be investing more work into patching the existing grant resources that have been the source of so much frustration and confusion. However, I will say that for your particular problem the root cause seems to be here:  https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/pkg/resources/grant_helpers.go#L271-L293. `enable_multiple_grants` is not being checked for shares, like it is for roles.

**Comment by Anatoli-deBRADKE** [comment]
Thanks for your answer @sfc-gh-swinkler  ! 

Is it possible to open a PR to "fix" enable_multiple_grants behaviour for shares? If it's the case, what is the process please? 

Thanks in advance!

**Comment by Anatoli-deBRADKE** [comment]
@sfc-gh-swinkler do you have any feedbacks on my previous message by any chance? Thanks in advance!

**Comment by scottwinkler** [comment]
fixed in pr #1510 . Will be released as part of 0.56.3

**Comment by Anatoli-deBRADKE** [comment]
Thanks a lot!
